session.go

```diff
@@ -1431,45 +1444,61 @@ func (s *session) shutdown() {
 	<-s.ctx.Done()
 }
 
-func (s *session) CloseWithError(code protocol.ApplicationErrorCode, desc string) error {
-	s.closeLocal(qerr.NewApplicationError(qerr.ErrorCode(code), desc))
+func (s *session) CloseWithError(code ApplicationErrorCode, desc string) error {
+	s.closeLocal(&qerr.ApplicationError{
+		ErrorCode:    code,
+		ErrorMessage: desc,
+	})
 	<-s.ctx.Done()
 	return nil
 }
 
-func (s *session) handleCloseError(closeErr closeError) {
-	if closeErr.err == nil {
-		closeErr.err = qerr.NewApplicationError(0, "")
+func (s *session) handleCloseError(closeErr *closeError) {
+	e := closeErr.err
+	if e == nil {
+		e = &qerr.ApplicationError{}
+	} else {
+		defer func() {
+			closeErr.err = e
+		}()
 	}
 
-	var quicErr *qerr.QuicError
-	var ok bool
-	if quicErr, ok = closeErr.err.(*qerr.QuicError); !ok {
-		quicErr = qerr.ToQuicError(closeErr.err)
+	if !errors.Is(e, qerr.ErrIdleTimeout) && !errors.Is(e, qerr.ErrHandshakeTimeout) &&
+		!errors.Is(e, &StatelessResetError{}) && !errors.Is(e, &VersionNegotiationError{}) &&
+		!errors.Is(e, &errCloseForRecreating{}) &&
+		!errors.Is(e, &qerr.ApplicationError{}) && !errors.Is(e, &qerr.TransportError{}) {
```

Comment:
🤢

Reply:
Hmm, can we not use a type switch here? I get that `As` is meant to peek at the error's layers, but I think `Is` isn't meant to do that - so just doing a type switch on the top-level error might be fine.

Reply:
A type switch with 7 empty cases and one default case that has some logic? Maybe that's at least a little bit nicer?

Reply:
Indeed, I'd find that nicer. You can even group the cases, like:

```
switch e.(type) {
case T1, T2, T3...:
default:
    // your logic
}
```

Reply:
> I get that As is meant to peek at the error's layers, but I think Is isn't meant to do that 

I think it is. From the docs: 
> Is reports whether any error in err's chain matches target.

Now we shouldn't encounter any wrapped errors here (at least not with the current code), but I guess it's still safer to use `errors.Is`.